pkg/flow/internal/controller/custom_component_registry.go

```diff
@@ -29,3 +34,40 @@ func (s *CustomComponentRegistry) registerDeclare(declare *ast.BlockStmt) {
 	defer s.mut.Unlock()
 	s.declares[declare.Label] = declare.Body
 }
+
+// registerImport stores the import namespace.
+// The content will be added later during evaluation.
+// It's important to register it before populating the component nodes
+// (else we don't know which one exists).
+func (s *CustomComponentRegistry) registerImport(importNamespace string) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	s.imports[importNamespace] = nil
+}
+
+// updateImportContent updates the content of a registered import.
+// The content of an import node can contain other import blocks.
+// These are considered as "children" of the root import node.
+// Each child has its own CustomComponentRegistry which needs to be updated.
+func (s *CustomComponentRegistry) updateImportContent(importNode *ImportConfigNode) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	if _, exist := s.imports[importNode.label]; !exist {
+		panic(fmt.Errorf("import %q was not registered", importNode.label))
+	}
+	importScope := NewCustomComponentRegistry(nil)
+	importScope.declares = importNode.ImportedDeclares()
+	importScope.updateImportContentChildren(importNode)
+	s.imports[importNode.label] = importScope
+}
+
+// updateImportContentChildren recurse through the children of an import node
+// and update their scope with the imported declare blocks.
+func (s *CustomComponentRegistry) updateImportContentChildren(importNode *ImportConfigNode) {
+	for _, child := range importNode.ImportConfigNodesChildren() {
+		childScope := NewCustomComponentRegistry(nil)
+		childScope.declares = child.ImportedDeclares()
+		childScope.updateImportContentChildren(child)
+		s.imports[child.label] = childScope
```

Comment:
Does `s.mut` need to be held here?

Reply:
updateImportContentChildren is meant to be only called by updateImportContent which is already holding the mut